src/cargo/core/compiler/fingerprint/mod.rs

```diff
@@ -772,10 +774,20 @@ impl LocalFingerprint {
     // TODO: This is allowed at this moment. Should figure out if it makes
     // sense if permitting to read env from the config system.
     #[allow(clippy::disallowed_methods)]
-    fn from_env<K: AsRef<str>>(key: K) -> LocalFingerprint {
+    fn from_env<K: AsRef<str>>(
+        key: K,
+        envs: &BTreeMap<String, Option<OsString>>,
+    ) -> LocalFingerprint {
         let key = key.as_ref();
         let var = key.to_owned();
-        let val = env::var(key).ok();
+        let val = envs
+            .get(key)
+            .and_then(|opt| {
+                opt.as_ref()
+                    .and_then(|os_str| os_str.clone().into_string().ok())
+            })
+            .or_else(|| env::var_os(key).and_then(|os_str| os_str.into_string().ok()));
```

Comment:
With using `env::var_os`, we can consolidate `.and_then(|os_str| os_str.into_string().ok())` to happen after both branches

```suggestion
        let val = envs
            .get(key)
            .map(|v| v.to_owned())
            .or_else(|| env::var_os(key))
            .and_then(|os_str| {
                os_str.into_string().ok()
            });
```

Reply:
thanks